soc/arm/st_stm32/stm32h7/mpu_regions.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2020 Mario Jaun
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <devicetree.h>
+#include <arch/arm/aarch32/cortex_m/mpu/arm_mpu.h>
+
+/* Flash Region Definitions */
+#if CONFIG_FLASH_SIZE <= 64
+#define REGION_FLASH_SIZE REGION_64K
+#elif CONFIG_FLASH_SIZE <= 128
+#define REGION_FLASH_SIZE REGION_128K
+#elif CONFIG_FLASH_SIZE <= 256
+#define REGION_FLASH_SIZE REGION_256K
+#elif CONFIG_FLASH_SIZE <= 512
+#define REGION_FLASH_SIZE REGION_512K
+#elif CONFIG_FLASH_SIZE <= 1024
+#define REGION_FLASH_SIZE REGION_1M
+#elif CONFIG_FLASH_SIZE <= 2048
+#define REGION_FLASH_SIZE REGION_2M
+#elif CONFIG_FLASH_SIZE <= 4096
+#define REGION_FLASH_SIZE REGION_4M
+#elif CONFIG_FLASH_SIZE <= 8192
+#define REGION_FLASH_SIZE REGION_8M
+#elif CONFIG_FLASH_SIZE <= 16384
+#define REGION_FLASH_SIZE REGION_16M
+#elif CONFIG_FLASH_SIZE <= 65536
+#define REGION_FLASH_SIZE REGION_64M
+#else
+#error "Unsupported flash size configuration"
+#endif
+
+/* SRAM Region Definitions */
+#if CONFIG_SRAM_SIZE <= 16
+#define REGION_SRAM_SIZE REGION_16K
+#elif CONFIG_SRAM_SIZE <= 32
+#define REGION_SRAM_SIZE REGION_32K
+#elif CONFIG_SRAM_SIZE <= 64
+#define REGION_SRAM_SIZE REGION_64K
+#elif CONFIG_SRAM_SIZE <= 128
+#define REGION_SRAM_SIZE REGION_128K
+#elif CONFIG_SRAM_SIZE <= 256
+#define REGION_SRAM_SIZE REGION_256K
+#elif CONFIG_SRAM_SIZE <= 512
+#define REGION_SRAM_SIZE REGION_512K
+#elif CONFIG_SRAM_SIZE <= 1024
+#define REGION_SRAM_SIZE REGION_1M
+#elif CONFIG_SRAM_SIZE <= 2048
+#define REGION_SRAM_SIZE REGION_2M
+#elif CONFIG_SRAM_SIZE <= 4096
+#define REGION_SRAM_SIZE REGION_4M
+#elif CONFIG_SRAM_SIZE == 32768
+#define REGION_SRAM_SIZE REGION_32M
+#else
+#error "Unsupported sram size configuration"
+#endif
+
+static const struct arm_mpu_region mpu_regions[] = {
+	MPU_REGION_ENTRY("FLASH", CONFIG_FLASH_BASE_ADDRESS,
+					 REGION_FLASH_ATTR(REGION_FLASH_SIZE)),
+	MPU_REGION_ENTRY("SRAM", CONFIG_SRAM_BASE_ADDRESS,
+					 REGION_RAM_ATTR(REGION_SRAM_SIZE)),
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay) && \
+		DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
+	MPU_REGION_ENTRY("SRAM3_ETH_BUF",
+					 DT_REG_ADDR(DT_NODELABEL(sram3)),
+					 REGION_NOCACHE_ATTR(REGION_32K)),
+	MPU_REGION_ENTRY("SRAM3_ETH_DESC",
+					 DT_REG_ADDR(DT_NODELABEL(sram3)),
+					 REGION_PPB_ATTR(REGION_256B)),
+#endif
```

Comment:
```suggestion
#if DT_NODE_HAS_STATUS(DT_NODELABEL(non_mpu_sram), okay) && \
		DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
	MPU_REGION_ENTRY("NON_MPU_SRAM_ETH_BUF",
					 DT_REG_ADDR(DT_NODELABEL(non_mpu_sram)),
					 REGION_NOCACHE_ATTR(REGION_SRAM_SIZE)),
	MPU_REGION_ENTRY("NON_MPU_SRAM_ETH_DESC",
					 DT_REG_ADDR(DT_NODELABEL(non_mpu_sram)),
					 REGION_PPB_ATTR(REGION_256B)),
#endif
```
Not sure if `REGION_SRAM_SIZE` correct here, maybe `REGION_8K` would be okay for the first time.

Reply:
Yes I think that should be `REGION_8K`, because `REGION_SRAM_SIZE` resolves to the size of the main system memory.

Reply:
To be honest, I think the name `non_mpu_sram` is a bit misleading, because the region uses a special MPU configuration, but the MPU is not disabled. For example the MPU still makes sure that the data in this region is not executable.

Reply:
Let stay with SRAM3